docs/source/explanations/dataclass.md

```diff
@@ -0,0 +1,455 @@
+
+```python
+from napari.utils.events.dataclass import evented_dataclass, Property
+
+@evented_dataclass(events=False, properties=True)
+class A:
+    x: int
+    y: Property[int, custom_getter, custom_setter] = 2
+    # where custom_getter/custom_setter are callables
+```
+
+Is *roughly* equivalent to:
+
+```python
+class A:
+    def __init__(self, x: int, y: int = 2):
+        self._x = x
+        self._y = y
+
+
+    @property
+    def x(self):
+        """Docs for "x" in the class docstring will be added here."""
+        return self._x
+
+    @x.setter
+    def x(self, value):
+        self._x = value
+
+    @property
+    def y(self):
+        """Docs for "y" in the class docstring will be added here."""
+        return custom_getter(self._y)
+
+    @y.setter
+    def y(self, value):
+        custom_setter(value)
+```
+
+## Examples and Special cases
+
+### Exempting a field from events
+
+Events will *only* be emitted on value change if the name of the attribute being
+set is one of the dataclass fields (i.e. ``name in self.__annotations__``), and
+the dataclass `__post_init__` method has already been called.
+
+In the following cases, events will *not* be emitted when the value of an
+attribute is set
+1. If the field is of type
+   [`dataclasses.InitVar`](https://docs.python.org/3/library/dataclasses.html#init-only-variables)
+2. If the field is of type
+   [`typing.ClassVar`](https://docs.python.org/3/library/dataclasses.html#class-variables)
+3. If the default value of the field is a call to
+   [`dataclasses.field`](https://docs.python.org/3/library/dataclasses.html#dataclasses.field)
+   with a `metadata` dict that contains `{'events': False}`
+
+```python
+from dataclasses import InitVar, field
+from typing import ClassVar
+
+from napari.utils.events.dataclass import evented_dataclass
+
+
+@evented_dataclass
+class Q:
+    # ONLY the `a` field will emit events
+    a: int = 0
+    b: str = field(default="hi", metadata={"events": False})
+    c: InitVar[int] = 0
+    d: ClassVar[float] = 1.0
+
+
+q = Q()
+print(set(q.events._emitters))  # prints: "{'a'}"
+```
+
+Note: event emission will also be skipped if the "new" value is the same as the
+existing value
+
+### Getter/setter side-effects & conditional event logic
+
+In some cases, it may be desireable to perform additional tasks or trigger
+side-effects upon setting an attribute (in addition to emitting the basic
+event). It may also be necessary at times to provide conditional logic that
+prevents the emission of an event in specific cases.  For both of these cases,
+you may optionally define **`_on_<name>_get`** and/or **`_on_<name>_set`**
+methods that will be called when getting/setting the corresponding `<name>`
+attribute.
```

Comment:
These get called before the actual thing is set/ gotten right, might be nice to add that to make it extra clear.

Reply:
Related question: are the `_on_{prop}_set()` methods called called to set the attribute values during the `__init__()`?